draft-ietf-mls-protocol.md

```diff
@@ -1435,44 +1465,230 @@ all arrive at the following state:
+                     |
+                     V
+           Derive-Secret(., "app upd","")
+                     |
+                     V
+           application_secret_N
+                     |
+                     +--> HKDF-Expand-Label(.,"nonce", "", nonce_length)
+                     |    = write_nonce_N_[sender]
+                     |
+                     +--> HKDF-Expand-Label(.,"key", "", key_length)
+                          = write_key_N_[sender]
+
+~~~~~
+
+The Application context provided together with the previous Application secret
+is used to bind the Application messages with the next key and add some freshness.
+
+[[OPEN ISSUE: Context is left empty for now, it might be that using only
+the message counter is enough, this would be more conveniant.
+Hashing all the data is obviously very costly and prevents from encrypt in
+parallel, an other solution could be to add a pseudo-random contribution to
+each message and only hash these. ]]
+
+[[ OPEN ISSUE: At the moment there is no contributivity of Application secrets
+chained from the initial one to the next generation of Epoch secret. While this
+seems safe because cryptographic operations using the application secrets can't
+affect the group init_secret, it remains to be proven correct. ]]
+
+### Updating the Application Secret
+
+When the Application Secret is updated:
+
+- Senders MUST use the generation N+1 of the application secret, where N is
+the last generation they received.
+- Recipients SHOULD delete older generations of application secret and as soon
+as possible, within usability bounds.
+
+These rules imply that in most circumstances, an application secret will be
+used for exactly one message. Recipients MUST NOT enforce
+a one-message-per-application-secret limit.
+
+### Application AEAD Key Calculation
 
-For every epoch, the root key of the ratcheting tree can be used to
-derive key material for symmetric operations such as encryption/AEAD and MAC;
-AEAD or MAC MUST be used to ensure that the message originated from a member
-of the group.
+The Application AEAD keying material is generated from the following
+input values:
 
-In addition, asymmetric signatures SHOULD be used to authenticate the sender
-of a message.
+- The Application Secret value;
+- A purpose value indicating the specific value being generated;
+- The length of the key being generated.
 
-In combination with server-side enforced ordering, data from previous messages
-is used (as a salt when hashing) to:
+The traffic keying material is generated from an input traffic secret value using:
 
- * add freshness to derived symmetric keys
- * cryptographically bind the transcript of all previous messages with the current
-   group shared secret
+~~~~
+write_nonce_[sender] =
+  HKDF-Expand-Label(Application_Secret,"mls app nonce", [sender], nonce_length)
 
-Possible candidates for that are:
+write_key_[sender] =
+  HKDF-Expand-Label(Application_Secret,"mls app key", [sender], key_length)
```

Comment:
This section is in conflict with the previous section.

Reply:
Fixed. Thanks !